goat/spider_utils.py

```diff
@@ -0,0 +1,315 @@
+from enum import Enum
+
+import requests
+from bs4 import BeautifulSoup
+
+
+class ExamType(str, Enum):
+    EGE = "EGE"
+    OGE = "OGE"
+    CT = "CT"
+    OLYMP = "OLYMP"
+
+
+class TaskType(str, Enum):
+    MULT_CHOICE = "MULT_CHOICE"
+    SOOTV = "SOOTV"
+    TEXT_ANSWER = "TEXT_ANSWER"
+
+
+class SdamgiaExamSubject(str, Enum):
+    MATH = "MATH"
+    MATHB = "MATHB"
+    PHYS = "PHYS"
+    INF = "INF"
+    RUS = "RUS"
+    BIO = "BIO"
+    EN = "EN"
+    CHEM = "CHEM"
+    GEO = "GEO"
+    SOC = "SOC"
+    DE = "DE"
+    FR = "FR"
+    LIT = "LIT"
+    SP = "SP"
+    HIST = "HIST"
+
+
+_BASE_DOMAIN = "sdamgia.ru"
+_SUBJECT_BASE_URL_ege = {
+    "MATH": f"https://math-ege.{_BASE_DOMAIN}",
+    "MATHB": f"https://mathb-ege.{_BASE_DOMAIN}",
+    "PHYS": f"https://phys-ege.{_BASE_DOMAIN}",
+    "INF": f"https://inf-ege.{_BASE_DOMAIN}",
+    "RUS": f"https://rus-ege.{_BASE_DOMAIN}",
+    "BIO": f"https://bio-ege.{_BASE_DOMAIN}",
+    "EN": f"https://en-ege.{_BASE_DOMAIN}",
+    "CHEM": f"https://chem-ege.{_BASE_DOMAIN}",
+    "GEO": f"https://geo-ege.{_BASE_DOMAIN}",
+    "SOC": f"https://soc-ege.{_BASE_DOMAIN}",
+    "DE": f"https://de-ege.{_BASE_DOMAIN}",
+    "FR": f"https://fr-ege.{_BASE_DOMAIN}",
+    "LIT": f"https://lit-ege.{_BASE_DOMAIN}",
+    "SP": f"https://sp-ege.{_BASE_DOMAIN}",
+    "HIST": f"https://hist-ege.{_BASE_DOMAIN}",
+}
+_SUBJECT_BASE_URL_oge = {
+    "MATH": f"https://math-oge.{_BASE_DOMAIN}",
+    "MATHB": f"https://mathb-oge.{_BASE_DOMAIN}",
+    "PHYS": f"https://phys-oge.{_BASE_DOMAIN}",
+    "INF": f"https://inf-oge.{_BASE_DOMAIN}",
+    "RUS": f"https://rus-oge.{_BASE_DOMAIN}",
+    "BIO": f"https://bio-oge.{_BASE_DOMAIN}",
+    "EN": f"https://en-oge.{_BASE_DOMAIN}",
+    "CHEM": f"https://chem-oge.{_BASE_DOMAIN}",
+    "GEO": f"https://geo-oge.{_BASE_DOMAIN}",
+    "SOC": f"https://soc-oge.{_BASE_DOMAIN}",
+    "DE": f"https://de-oge.{_BASE_DOMAIN}",
+    "FR": f"https://fr-oge.{_BASE_DOMAIN}",
+    "LIT": f"https://lit-oge.{_BASE_DOMAIN}",
+    "SP": f"https://sp-oge.{_BASE_DOMAIN}",
+    "HIST": f"https://hist-oge.{_BASE_DOMAIN}",
+}
+_RESHU_CT = "reshuct.by"
+_SUBJECT_BASE_URL_ct = {
+    "MATH": f"https://math3.{_RESHU_CT}",
+    "MATHB": f"https://math3b.{_RESHU_CT}",
+    "PHYS": f"https://phys.{_RESHU_CT}",
+    "INF": f"https://inf.{_RESHU_CT}",
+    "RUS": f"https://rus.{_RESHU_CT}",
+    "BIO": f"https://bio.{_RESHU_CT}",
+    "EN": f"https://en.{_RESHU_CT}",
+    "CHEM": f"https://chem.{_RESHU_CT}",
+    "GEO": f"https://geo.{_RESHU_CT}",
+    "SOC": f"https://soc.{_RESHU_CT}",
+    "DE": f"https://de.{_RESHU_CT}",
+    "FR": f"https://fr.{_RESHU_CT}",
+    "LIT": f"https://lit.{_RESHU_CT}",
+    "SP": f"https://sp.{_RESHU_CT}",
+    "WH": f"https://wh.{_RESHU_CT}",
+    "BH": f"https://bh.{_RESHU_CT}",
+}
+
+
+def determine_soc_task_type(exam_type: str, topic_id: str) -> tuple[TaskType, bool]:
```

Comment:
You can do a little refactoring here and pull all these constants into a separate mapping and check the type inside the loop. This is not necessary, it's a matter of code beautification.